core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -44,6 +44,12 @@ class PartitionStateMachine(controller: KafkaController) extends Logging {
   private val partitionState: mutable.Map[TopicAndPartition, PartitionState] = mutable.Map.empty
   private val brokerRequestBatch = new ControllerBrokerRequestBatch(controller)
   private val noOpPartitionLeaderSelector = new NoOpLeaderSelector(controllerContext)
+  private val validPreviousStates: Map[PartitionState, Set[PartitionState]] =
+    Map(NewPartition -> Set(NonExistentPartition),
+      OnlinePartition -> Set(NewPartition, OnlinePartition, OfflinePartition),
+      OfflinePartition -> Set(NewPartition, OnlinePartition, OfflinePartition),
+      NonExistentPartition -> Set(OfflinePartition))
```

Comment:
Did you consider defining this in the `PartitionState` types? For example:

```scala
case object NewPartition extends PartitionState {
  val state: Byte = 0
  val validPreviousStates = Set(NonExistentPartition)
}
```
If a new state is added, it will be clear (i.e. there will be a compiler error) that this method needs to be implemented.